File: 프로그래머스/알고리즘 고득점 Kit/깊이,너비 우선 탐색(DFS,BFS)/Level 2/타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

int DFS(vector<int> numbers, int target, int index, int n){
    if(index >= numbers.size())
        return n==target ? 1 : 0;
    
    int cnt = 0;
    cnt += DFS(numbers, target, index+1, n+numbers[index]);
    cnt += DFS(numbers, target, index+1, n-numbers[index]);
    
    return cnt;
}

int solution(vector<int> numbers, int target) {
    return DFS(numbers, target, 0, 0);
}

File: 프로그래머스/알고리즘 고득점 Kit/깊이,너비 우선 탐색(DFS,BFS)/Level 2/타겟 넘버/타겟 넘버.py
def DFS(numbers, target, index, n):
    if index >= len(numbers):
        if n == target: return 1
        else: return 0

    cnt = 0
    cnt += DFS(numbers, target, index+1, n+numbers[index])
    cnt += DFS(numbers, target, index+1, n-numbers[index])

    return cnt

def solution(numbers, target):
    return DFS(numbers, target, 0, 0)
